pkg/workspace/resource/node_claim.go

```diff
@@ -47,31 +50,54 @@ func NewNodeClaimManager(c client.Client, recorder record.EventRecorder, expecta
 	}
 }
 
-// CheckNodeClaims checks the current state of NodeClaims with the target count recorded in the workspace.Status.
-func (c *NodeClaimManager) CheckNodeClaims(ctx context.Context, wObj *kaitov1beta1.Workspace) (int, []*karpenterv1.NodeClaim, []string, error) {
-	var addedNodeClaimsCount int
+func (c *NodeClaimManager) GetNumNodeClaimsNeeded(ctx context.Context, wObj *kaitov1beta1.Workspace, readyNodes []*corev1.Node) int {
+	targetNodeCount := int(wObj.Status.TargetNodeCount)
 
-	// Calculate the number of NodeClaims required (target - BYO nodes)
-	readyNodes, targetNodeClaimsCount, err := nodeclaim.ResolveReadyNodesAndTargetNodeClaimCount(ctx, c.Client, wObj)
-	if err != nil {
-		return addedNodeClaimsCount, nil, nil, fmt.Errorf("failed to get required NodeClaims: %w", err)
+	// Count ready nodes that do NOT have a corresponding NodeClaim
+	readyNodesWithoutNodeClaim := 0
+
+	for _, node := range readyNodes {
+		if _, ok := node.Labels[consts.LabelNodePool]; !ok {
+			readyNodesWithoutNodeClaim++
+		}
 	}
 
-	ncList, err := nodeclaim.ListNodeClaim(ctx, wObj, c.Client)
+	// Calculate how many NodeClaims we need, including those already provisioned.
+	targetNodeClaimCount := max(0, targetNodeCount-readyNodesWithoutNodeClaim)
```

Comment:
I want to check two points for calculating targetNodeClaimCount:
1. We account on the number of ready BYO nodes to calculate targetNodeClaimCount, this means if one BYO node falls into not ready, a new NodeClaim will be created to replace it. and after some time, If the BYO node comes back to ready, the created NodeClaim will be leaked because it seems that we have not considered NodeClaim deletion logic.
2. If our design is: one BYO node falls into not ready, a new NodeClaim will be created to replace it. There is a problem is `workspace controller` has not watched nodes, so workspace reconcile will not be triggered when BYO node becomes not ready.